### ISSUE TITLE
False positive: Redundant cast to string

### ISSUE THREAD

**Issue by BenMorel** [vimeo/psalm#4891]
An object can have numeric dynamic properties, in which case `ReflectionProperty::getName()`, when used as an array key, may be cast to an integer.

Demonstration:

```php
$o = new stdClass;
$o->{'123'} = 'hello';

$r = new ReflectionObject($o);

$props = [];

foreach ($r->getProperties() as $p) {
    $name = $p->getName();
    $props[$name] = 'x';
}

foreach ($props as $key => $value) {
    echo gettype($key); // integer
    
    $key = (string) $key; // ERROR: RedundantCast - 18:12 - Redundant cast to string
    
    echo "$key = $value";
}
```

https://psalm.dev/r/5442f2cb5f

**Comment by orklah** [comment]
The underlying issue is basically https://psalm.dev/r/7d3fb8833d (except when the numeric-string is exotic)

I intend to push a PR in the near future to allow Psalm distinguish when a given type comes from an array offset to help with that. If it's accepted, the RedundantCast will be easy to solve

**Comment by BenMorel** [comment]
The issue with `numeric-string`, IMO, is that it's too broad and cannot accurately determine if a string may be cast to an integer when used as an array key.

For example, `'123'` and `'-12'` will be cast to `int`, while `'1.23'`, `'012'` or `'+12'` will stay as strings, even though all of them are `numeric-string`s.

For this reason, I think it would be interesting to introduce a couple additional types in Psalm: `int-string` and `non-int-string`. I would define `int-string` as a string that matches `^(0|\-?[1-9][0-9]*)$`, whatever the length.

As I see it, an array that's documented as `array<string, T>` should be considered internally as `array<int|string, T>`, because there's no guarantee that keys will not be silently converted to integers. 

On the other hand, if you document it as `array<non-int-string, T>`, *then* we can safely assume that keys will always be strings.

As an aside, I would make heavy use of an `int-string` type in [brick/math](https://github.com/brick/math/blob/242b0ec0f3532491f922950e72602816398b62a0/src/BigInteger.php#L24-L32)!
